modular_zubbers/code/modules/storyteller/storyteller.dm

```diff
@@ -0,0 +1,140 @@
+	var/guarantees_roundstart_roleset = TRUE
+
+	/// Whether the storyteller has the distributions disabled. Important for ghost storytellers
+	var/disable_distribution = FALSE
+
+	/// Whether people can vote for the storyteller
+	var/votable = TRUE
+	/// If defined, will need a minimum of population to be votable
+	var/population_min
+	/// If defined, it will not be votable if exceeding the population
+	var/population_max
+
+/datum/storyteller/process(delta_time)
+	if(disable_distribution)
+		return
+	add_points(delta_time)
+	handle_tracks()
+
+/// Add points to all tracks while respecting the multipliers.
+/datum/storyteller/proc/add_points(delta_time)
+	var/datum/controller/subsystem/gamemode/mode = SSgamemode
+	var/base_point = EVENT_POINT_GAINED_PER_SECOND * delta_time * mode.event_frequency_multiplier
+	for(var/track in mode.event_track_points)
+		var/point_gain = base_point * point_gains_multipliers[track] * mode.point_gain_multipliers[track]
+		if(mode.allow_pop_scaling)
+			point_gain *= mode.current_pop_scale_multipliers[track]
+		mode.event_track_points[track] += point_gain
+		mode.last_point_gains[track] = point_gain
+
+/// Goes through every track of the gamemode and checks if it passes a threshold to buy an event, if does, buys one.
+/datum/storyteller/proc/handle_tracks()
+	. = FALSE //Has return value for the roundstart loop
+	var/datum/controller/subsystem/gamemode/mode = SSgamemode
+	for(var/track in mode.event_track_points)
+		var/points = mode.event_track_points[track]
+		if(points >= mode.point_thresholds[track] && find_and_buy_event_from_track(track))
+			. = TRUE
+
+/// Find and buy a valid event from a track.
+/datum/storyteller/proc/find_and_buy_event_from_track(track)
+	. = FALSE
+	var/datum/controller/subsystem/gamemode/mode = SSgamemode
+	var/datum/round_event_control/picked_event
+	if(mode.forced_next_events[track]) //Forced event by admin
+		/// Dont check any prerequisites, it has been forced by an admin
+		picked_event = mode.forced_next_events[track]
+		mode.forced_next_events -= track
+	else
+		mode.update_crew_infos()
+		var/pop_required = mode.min_pop_thresholds[track]
+		if(mode.active_players < pop_required)
+			message_admins("Storyteller failed to pick an event for track of [track] due to insufficient population. (required: [pop_required] active pop for [track]. Current: [mode.active_players])")
+			mode.event_track_points[track] *= TRACK_FAIL_POINT_PENALTY_MULTIPLIER
+			return
+		calculate_weights(track)
+		var/list/valid_events = list()
+		// Determine which events are valid to pick
+		for(var/datum/round_event_control/event as anything in mode.event_pools[track])
+			if(event.can_spawn_event())
+				valid_events[event] = event.calculated_weight
+		///If we didn't get any events, remove the points inform admins and dont do anything
+		if(!length(valid_events))
+			message_admins("Storyteller failed to pick an event for track of [track].")
```

Comment:
needs to be logged too, in a separate storyteller logfile ideally. Or the dynamic logfile